[FLINK-28162][sql-gateway]Introduce the operation related API for REST endpoint

WencongLiu opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

*Introduce the operation related api in sql gateway*


## Brief change log

  - *Introduce operation related handlers and message headers*


## Verifying this change

This change added tests and can be verified as follows:

  - *Added integration tests for operation related apis*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (not documented)

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ExceptionInfo.java:
```diff
@@ -0,0 +1,64 @@
+/** ExceptionInfo for SQL gateway service. */
+public class ExceptionInfo {
+
+    private static final String FIELD_EXCEPTION_ROOT_CLASS_NAME = "root";
```
Should be "root_cause" as specified in FLIP.

WencongLiu replied:
Fixed.
The ExceptionInfo contains the same attributes as the FLIP.

"exception": {

      "root_cause": "....",

     "exception_stack": "..." 

}

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ExceptionInfo.java:
```diff
@@ -0,0 +1,64 @@
+public class ExceptionInfo {
+
+    private static final String FIELD_EXCEPTION_ROOT_CLASS_NAME = "root";
+    private static final String FIELD_EXCEPTION_MESSAGE = "message";
```
This is not specified in FLIP.

WencongLiu replied:
Fixed.
The ExceptionInfo contains the same attributes as the FLIP.

"exception": {

      "rootCause": "....",

     "exceptionStack": "..." 

}

xintongsong commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/FetchResultsResponseBody.java:
```diff
@@ -0,0 +1,75 @@
+    @JsonProperty(FIELD_RESULT_TYPE)
+    private final String resultType;
+
+    @JsonProperty(FIELD_NEXT_RESULT_URI)
+    private final String nextResultUri;
+
+    @JsonProperty(FIELD_EXCEPTION)
+    private final ExceptionInfo exceptionInfo;
```
IIUC, these two should be nullable?

WencongLiu replied:
These two fields should not be nullable.
If there is no more data,  `nextResultUri ` should be an empty string `""`.
If there is no exception throwed, all fields in 'exceptionInfo' should be null.
The content of these two field in the  response will be 

"next_result_uri": "".
    "exception": {
      "root_cause": null,
      "exception_stack": null 
   }

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSet.java:
```diff
@@ -23,16 +23,30 @@
+@JsonDeserialize(using = ResultSetJsonDeserializer.class)
+public class ResultSet implements Serializable {
+
+    public static final String FIELD_NAME_RESULT_TYPE = "result_type";
```
This is not defined in FLIP

WencongLiu replied:
Removed.

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSet.java:
```diff
@@ -23,16 +23,30 @@
+
+    public static final String FIELD_NAME_DATA = "data";
+
+    public static final String FIELD_NAME_NEXT_TOKEN = "next_token";
```
This is not defined in FLIP

WencongLiu replied:
Removed.

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSet.java:
```diff
@@ -23,16 +23,30 @@
-public class ResultSet {
+@JsonSerialize(using = ResultSetJsonSerializer.class)
+@JsonDeserialize(using = ResultSetJsonDeserializer.class)
+public class ResultSet implements Serializable {
```
It seems this class needs a serial version id

WencongLiu replied:
Fixed.

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSetJsonSerializer.java:
```diff
@@ -0,0 +1,89 @@
+            throws IOException {
+        jsonGenerator.writeStartObject();
+
+        // Serialize result type
+        serializerProvider.defaultSerializeField(
+                FIELD_NAME_RESULT_TYPE, resultSet.getResultType(), jsonGenerator);
+
+        // Serialize next token
+        serializerProvider.defaultSerializeField(
+                FIELD_NAME_NEXT_TOKEN, resultSet.getNextToken(), jsonGenerator);
```
These are not defined in FLIP.

WencongLiu replied:
Removed.

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSetJsonSerializer.java:
```diff
@@ -0,0 +1,89 @@
+                        JsonFormatOptions.MapNullKeyMode.LITERAL,
+                        "null",
+                        true);
+        List<String> rowDataStrings = new ArrayList<>();
+        for (RowData row : resultSet.getData()) {
+            byte[] rowByte = serializationSchema.serialize(row);
+            rowDataStrings.add(new String(rowByte, StandardCharsets.UTF_8));
+        }
```
These doesn't align with the FLIP.

WencongLiu replied:
Fixed. 
Now `data` contains a list of Json Objects. Each of Json Object contains a RowKind and a list of JsonNode.

xintongsong commented on flink-table/flink-sql-gateway-api/pom.xml:
```diff
@@ -64,7 +64,19 @@
+        <dependency>
+            <groupId>org.apache.flink</groupId>
+            <artifactId>flink-json</artifactId>
+            <version>1.16-SNAPSHOT</version>
```
```suggestion
            <version>${project.version}</version>
```

WencongLiu replied:
Fixed.

xintongsong commented on flink-table/flink-sql-gateway-api/pom.xml:
```diff
@@ -64,7 +64,19 @@
+		<dependency>
+			<groupId>org.apache.flink</groupId>
+			<artifactId>flink-table-runtime</artifactId>
+			<version>1.16-SNAPSHOT</version>
```
```suggestion
			<version>${project.version}</version>
```

WencongLiu replied:
Fixed.

xintongsong commented on flink-table/flink-sql-gateway-api/src/main/java/org/apache/flink/table/gateway/api/results/ResultSetJsonSerializer.java:
```diff
@@ -0,0 +1,89 @@
+import static org.apache.flink.table.gateway.api.results.ResultSet.FIELD_NAME_RESULT_TYPE;
+
+/** Json serializer for {@link ResultSet}. */
+public class ResultSetJsonSerializer extends JsonSerializer<ResultSet> {
```
We need tests for this serializer and the corresponding deserializer.

WencongLiu replied:
Added tests.

zentol commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/FetchResultsRequestBody.java:
```diff
@@ -0,0 +1,47 @@
+@JsonInclude(JsonInclude.Include.NON_NULL)
+public class FetchResultsRequestBody implements RequestBody {
+
+    private static final String FIELD_NAME_MAX_FETCH_SIZE = "max_fetch_size";
```
Please make sure that _all_ request/response body fields use camelCase.

xintongsong replied:
I agree that it would be nice to use consistent naming conventions for the rest field in 2.0, and we should start to do that early. Is there already a consensus that camelCase will be used in long term? Could you point us to it?

zentol replied:
See the discussion in https://issues.apache.org/jira/browse/FLINK-15787

I also made sure all recently added REST APIs followed camelCase.

xintongsong replied:
@zentol, thanks a lot for the pointer.

@WencongLiu, let's also add a hotfix commit and fix this for the APIs we added in FLINK-28161.

WencongLiu replied:
@zentol @xintongsong thanks, I'll fix it.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/operation/AbstractOperationHandler.java:
```diff
@@ -0,0 +1,91 @@
+        }
+    }
+
+    boolean execute(SessionHandle sessionHandle, OperationHandle operationHandle) {
```
The class inheritance structure has been refactored.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/util/GetApiVersionHandler.java:
```diff
@@ -0,0 +1,58 @@
+    protected CompletableFuture<GetApiVersionResponseBody> handleRequest(
+            @Nullable SqlGatewayRestAPIVersion version,
+            @Nonnull HandlerRequest<EmptyRequestBody> request) {
+        return CompletableFuture.completedFuture(
+                new GetApiVersionResponseBody(
+                        Collections.singletonList(SqlGatewayRestAPIVersion.V1.toString())));
```
Done.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/statement/ExecuteStatementHandler.java:
```diff
@@ -0,0 +1,89 @@
+
+package org.apache.flink.table.gateway.rest.handler.statement;
+
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
```
Removed.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/ExecuteStatementRequestBody.java:
```diff
@@ -0,0 +1,59 @@
+    private static final String FIELD_NAME_EXECUTION_TIMEOUT = "execution_timeout";
+
+    @JsonProperty(FIELD_NAME_STATEMENT)
+    @Nullable
+    private final String statement;
```
Fixed.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/ExecuteStatementResponseBody.java:
```diff
@@ -0,0 +1,53 @@
+    private final String operationHandle;
+
+    @JsonProperty(FIELD_OPERATION_TYPE)
+    private final String operationTYpe = OperationType.EXECUTE_STATEMENT.name();
```
Fixed.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/ExecuteStatementResponseBody.java:
```diff
@@ -0,0 +1,53 @@
+        return operationHandle;
+    }
+
+    public String getOperationTYpe() {
```
Fixed.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/ExecuteStatementResponseBody.java:
```diff
@@ -0,0 +1,53 @@
+    @JsonProperty(FIELD_OPERATION_TYPE)
+    private final String operationTYpe = OperationType.EXECUTE_STATEMENT.name();
+
+    public ExecuteStatementResponseBody(
```
The field `has_result` will be removed in FLIP.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/statement/ExecuteStatementHandler.java:
```diff
@@ -0,0 +1,89 @@
+        timeout = timeout == null ? 0L : timeout;
+        SessionHandle sessionHandle = request.getPathParameter(SessionHandleIdPathParameter.class);
+        OperationHandle operationHandle =
+                service.executeStatement(sessionHandle, statement, timeout, new Configuration());
```
Have add a `Map<String, String>` in `ExecuteStatementResponseBody` as the value of executionConfig.

WencongLiu commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/FetchResultsRequestBody.java:
```diff
@@ -0,0 +1,47 @@
+
+    private static final String FIELD_NAME_MAX_FETCH_SIZE = "max_fetch_size";
+
+    @JsonProperty(FIELD_NAME_MAX_FETCH_SIZE)
+    @Nullable
+    private final Integer maxFetchSize;
```
This field has been deleted.

WencongLiu commented:
@flinkbot run azure

xintongsong reviewed (commented):
Thanks @WencongLiu. I think this PR is very close to a mergable state.

I have rebased the latest master branch, squashed your fixup commits, and made a few minor changes in two new fixup commits. I've also moved the session related api changes to a hotfix commit.

I have only two remaining issues.
- I left some of the comments from the previous round of review unresolved, most of which requires updates to the FLIP. @fsk119, could you help confirm them?
- I'm not fully convinced by making `nextResultUri` and `exception` in `FetchResultsResponseBody` required, even when there's not a next result or exception, especially for `exception`. Why do we need this complexity?

WencongLiu commented:
> Thanks @WencongLiu. I think this PR is very close to a mergable state.
> 
> I have rebased the latest master branch, squashed your fixup commits, and made a few minor changes in two new fixup commits. I've also moved the session related api changes to a hotfix commit.
> 
> I have only two remaining issues.
> 
> * I left some of the comments from the previous round of review unresolved, most of which requires updates to the FLIP. @fsk119, could you help confirm them?
> * I'm not fully convinced by making `nextResultUri` and `exception` in `FetchResultsResponseBody` required, even when there's not a next result or exception, especially for `exception`. Why do we need this complexity?

xintongsong commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/header/statement/FetchResultsHeaders.java:
```diff
@@ -0,0 +1,97 @@
+        return INSTANCE;
+    }
+
+    public static String buildNextUri(
```
```suggestion
    @Nullable
    public static String buildNextUri(
```

xintongsong commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/message/statement/FetchResultsResponseBody.java:
```diff
@@ -0,0 +1,67 @@
+    private static final String FIELD_RESULT_TYPE = "resultType";
+    private static final String FIELD_RESULTS = "results";
+    private static final String FIELD_NEXT_RESULT_URI = "nextResultUri";
+    private static final String FIELD_EXCEPTION = "exception";
```
```suggestion
```

xintongsong commented on flink-table/flink-sql-gateway/src/main/java/org/apache/flink/table/gateway/rest/handler/statement/FetchResultsHandler.java:
```diff
@@ -0,0 +1,97 @@
+            nextToken = resultSet.getNextToken();
+            resultType = resultSet.getResultType().toString();
+        } catch (Exception e) {
+            Throwable root = e;
+            while (root.getCause() != null
+                    && root.getCause().getMessage() != null
+                    && !root.getCause().getMessage().isEmpty()) {
+                root = root.getCause();
+            }
+            throw new SqlGatewayException(root);
```
I think we should preserve the complete error stack.
```suggestion
            throw new SqlGatewayException(e);
```

xintongsong reviewed (commented):
Thanks @WencongLiu. I have a few more comments.

xintongsong reviewed (approved):
Thanks @WencongLiu. LGTM. I'll merge this upon the CI success.

WencongLiu commented:
@flinkbot run azure

xintongsong commented:
@flinkbot run azure
